homeassistant/components/generic_thermostat/climate.py

```diff
@@ -132,336 +333,529 @@ class GenericThermostat(ClimateDevice, RestoreEntity):
+                    else STATE_OFF
+                )
+                if self._hvac_mode == HVAC_MODE_COOL:
+                    entity_id = self._ac_entity_id
+                    current_state = (
+                        STATE_ON
+                        if self._is_cool_enabled and self._is_ac_active
+                        else STATE_OFF
                     )
-                    if not long_enough:
-                        return
-
-            too_cold = self._target_temp >= self._cur_temp + self._cold_tolerance
-            too_hot = self._cur_temp >= self._target_temp + self._hot_tolerance
-            if self._is_device_active:
-                if (self.ac_mode and too_cold) or (not self.ac_mode and too_hot):
-                    _LOGGER.info("Turning off heater %s", self.heater_entity_id)
-                    await self._async_heater_turn_off()
-                elif time is not None:
-                    # The time argument is passed only in keep-alive case
-                    _LOGGER.info(
-                        "Keep-alive - Turning on heater heater %s",
-                        self.heater_entity_id,
+
+                long_enough = condition.state(
+                    self.hass, entity_id, current_state, self._min_cycle_duration
+                )
+
+                if not long_enough:
+                    _LOGGER.debug(
+                        "Operate - Min duration not expired, exiting (%s, %s, %s)",
+                        self._min_cycle_duration,
+                        current_state,
+                        entity_id,
                     )
-                    await self._async_heater_turn_on()
+                    return
+
+            target_temp_min = self.target_temperature  # lower limit
+            target_temp_max = self.target_temperature  # upper limit
+            if self._hvac_mode == HVAC_MODE_HEAT:
+                target_temp_min = target_temp_min - self._tolerance_on
+                target_temp_max = target_temp_max + self._tolerance_off
             else:
-                if (self.ac_mode and too_hot) or (not self.ac_mode and too_cold):
-                    _LOGGER.info("Turning on heater %s", self.heater_entity_id)
-                    await self._async_heater_turn_on()
-                elif time is not None:
-                    # The time argument is passed only in keep-alive case
-                    _LOGGER.info(
-                        "Keep-alive - Turning off heater %s", self.heater_entity_id
-                    )
-                    await self._async_heater_turn_off()
+                target_temp_min = target_temp_min - self._tolerance_off
+                target_temp_max = target_temp_max + self._tolerance_on
+
+            current_temp = self._current_temperature
+
+            _LOGGER.debug(
+                "Operate - target_temp_min %s, target_temp_max %s, current_temp %s, target_temp %s, shift %s, keepalive %s",
```

Comment:
Please break long strings around 88 characters.

Reply:
I've break the line